File: rock/ts-sdk/src/job/operator.ts
/**
 * Operator — generic algorithm that produces a TrialList from a JobConfig.
 *
 * Matches Python rock.sdk.job.operator.
 */

import type { AbstractTrial } from './trial/abstract';
import { createTrial } from './trial/registry';

// ---------------------------------------------------------------------------
// Operator interface
// ---------------------------------------------------------------------------

/**
 * Operator base: apply(config) -> list[AbstractTrial].
 *
 * Operators generate a TrialList from a config. They don't manage
 * sandbox lifecycle (JobExecutor does) — just decide what to run.
 */
export interface Operator {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  apply(config: Record<string, any>): AbstractTrial[];
}

// ---------------------------------------------------------------------------
// ScatterOperator
// ---------------------------------------------------------------------------

/**
 * Scatter: create `size` identical Trial instances from config.
 *
 * Analog of torch.distributed.scatter — same data/config distributed to N workers.
 *
 * Usage:
 *   new ScatterOperator()      // size=1, single trial (default)
 *   new ScatterOperator(8)     // 8 parallel trials
 *   new ScatterOperator(0)     // empty list, no-op
 */
export class ScatterOperator implements Operator {
  readonly size: number;

  constructor(size: number = 1) {
    this.size = size;
  }

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  apply(config: Record<string, any>): AbstractTrial[] {
    if (this.size <= 0) return [];
    const trial = createTrial(config);
    return Array(this.size).fill(trial);
  }
}

File: rock/sdk/job/operator.py
"""Operator primitives that produce Trial lists from a JobConfig."""

from __future__ import annotations

from abc import ABC, abstractmethod
from typing import TYPE_CHECKING

from rock.sdk.job.trial.registry import _create_trial

if TYPE_CHECKING:
    from rock.sdk.job.config import JobConfig
    from rock.sdk.job.trial.abstract import AbstractTrial


class Operator(ABC):
    """Operator base: apply(config) -> list[AbstractTrial]."""

    @abstractmethod
    def apply(self, config: JobConfig) -> list[AbstractTrial]:
        """Generate a TrialList from config. Empty list means no-op."""
        ...


class ScatterOperator(Operator):
    """Create ``size`` identical Trial instances from config."""

    def __init__(self, size: int = 1):
        self.size = size

    def apply(self, config: JobConfig) -> list[AbstractTrial]:
        if self.size <= 0:
            return []
        trial = _create_trial(config)
        return [trial] * self.size
